lost/src/main/java/com/mapzen/android/lost/internal/GeofencingApiImpl.java

```diff
@@ -16,32 +19,63 @@
+    pendingIntentMap = new HashMap<>();
   }
 
+  @RequiresPermission(anyOf = {Manifest.permission.ACCESS_COARSE_LOCATION, Manifest.permission.ACCESS_FINE_LOCATION})
   @Override
   public void addGeofences(GeofencingRequest geofencingRequest, PendingIntent pendingIntent)
       throws SecurityException {
-    ParcelableGeofence geofence = (ParcelableGeofence) geofencingRequest.getGeofences().get(0);
-    locationManager.addProximityAlert(
-        geofence.getLatitude(),
-        geofence.getLongitude(),
-        geofence.getRadius(),
-        geofence.getDuration(),
-        pendingIntent);
+    List<Geofence> geofences = geofencingRequest.getGeofences();
+    addGeofences(geofences, pendingIntent);
   }
 
-  @Override public void addGeofences(List<Geofence> geofences, PendingIntent pendingIntent) {
-    throw new RuntimeException("Sorry, not yet implemented");
+  @RequiresPermission(anyOf = {Manifest.permission.ACCESS_COARSE_LOCATION, Manifest.permission.ACCESS_FINE_LOCATION})
+  @Override public void addGeofences(List<Geofence> geofences, PendingIntent pendingIntent)
+      throws SecurityException {
+    for (Geofence geofence : geofences) {
+      addGeofence(geofence, pendingIntent);
+    }
   }
 
+  @RequiresPermission(anyOf = {Manifest.permission.ACCESS_COARSE_LOCATION, Manifest.permission.ACCESS_FINE_LOCATION})
+  private void addGeofence(Geofence geofence, PendingIntent pendingIntent)
+          throws SecurityException {
+    ParcelableGeofence pGeofence = (ParcelableGeofence) geofence;
+    String requestId = String.valueOf(pGeofence.hashCode());
+    locationManager.addProximityAlert(
+            pGeofence.getLatitude(),
+            pGeofence.getLongitude(),
+            pGeofence.getRadius(),
+            pGeofence.getDuration(),
+            pendingIntent);
+    if(pGeofence.getRequestId()!=null && !pGeofence.getRequestId().isEmpty()){
+      requestId = pGeofence.getRequestId();
+    }
+    pendingIntentMap.put(requestId, pendingIntent);
+  }
+
+  @RequiresPermission(anyOf = {Manifest.permission.ACCESS_COARSE_LOCATION, Manifest.permission.ACCESS_FINE_LOCATION})
   @Override public void removeGeofences(List<String> geofenceRequestIds) {
-    throw new RuntimeException("Sorry, not yet implemented");
+    for (String geofenceRequestId : geofenceRequestIds) {
+      removeGeofences(geofenceRequestId);
+    }
+  }
+
+  @RequiresPermission(anyOf = {Manifest.permission.ACCESS_COARSE_LOCATION, Manifest.permission.ACCESS_FINE_LOCATION})
+  private void removeGeofences(String geofenceRequestId) {
+    PendingIntent pendingIntent = pendingIntentMap.get(geofenceRequestId);
+    removeGeofences(pendingIntent);
   }
 
-  @Override public void removeGeofences(PendingIntent pendingIntent) {
-    throw new RuntimeException("Sorry, not yet implemented");
+  @RequiresPermission(anyOf = {Manifest.permission.ACCESS_COARSE_LOCATION, Manifest.permission.ACCESS_FINE_LOCATION})
+  @Override public void removeGeofences(PendingIntent pendingIntent)
```

Comment:
We could also add something like this to `GeofencingApiTest`:

```java
@Test public void removeGeofence_pendingIntent_shouldRemoveProximityAlert() {
    PendingIntent intent = Mockito.mock(PendingIntent.class);
    geofencingApi.removeGeofences(intent);
    Mockito.verify(locationManager, times(1)).removeProximityAlert(intent);
  }
```
And then we would just be missing a test for `GeofencingApiImpl#removeGeofences(String)`...but I imagine it would look similar to the above. Do you want to add these tests?